stdlib/std.jsonnet

```diff
@@ -1599,20 +1599,35 @@ limitations under the License.
     if !std.isArray(arr) then
       error 'all() parameter should be an array, got ' + std.type(arr)
     else
-      local and = function(x, y)
-        if !std.isBoolean(y) then
-          error std.format('element "%s" of type %s is not a boolean', y, std.type(y))
-        else x && y;
-      std.foldl(and, arr, true),
+      local arrLen = std.length(arr);
+      local aux(idx) =
+        local e = arr[idx];
+        if idx >= arrLen then
```

Comment:
What happens to that array index operation when the supplied index exceeds its length minus one?

Reply:
While not obvious, and therefore could be addressed, this doesn't cause a runtime issue. Since Jsonnet is evaluated lazily, and since `e` isn't used until after the index check, it won't throw an exception:

```shell
$ echo 'std.any([])' | bazel run -c opt //cmd:jsonnet 2> /dev/null -- -
false

$ echo 'std.all([])' | bazel run -c opt //cmd:jsonnet 2> /dev/null -- -
true
```

Reply:
That was definitely not obvious to me, just reading the code. What led you to hoist that index operation up like that? Consider this slight revision:
```jsonnet
local aux(idx) =
  if idx >= arrLen then
    true
  else
    local e = arr[idx];
    if !std.isBoolean(e) then
      error std.format('element "%s" of type %s is not a boolean', e, std.type(e))
    else if !e then
      false
    else
      aux(idx + 1) tailstrict;
```

Reply:
I don't mind changing to make it more clear. I will amend the PR later.

As for why I made the choice, I probably didn't notice the array index check. And I was further biased by the unit tests passing. 🙃